More improvements (#10)
Lawrencium77/quant_blog: Fast and Accurate GPU Quantization for Transformers.md

# Part II: Fast GPU Quantization in Practice

In order to run INT8 GEMMs efficiently on CUDA GPUs we must execute the operation against INT8 Tensor Cores. These were first introduced with the Turing architecture (compute capability 7.0+). INT4 and INT1 Tensor Cores also exist but have been deprecated in future architectures (see the figure below, taken from [Wikipedia](https://en.wikipedia.org/wiki/Ampere_(microarchitecture))). We therefore focus on INT8 quantization.

![](_attachments/Pasted%20image%2020230416203418.png)

Executing against Tensor Cores can be achieved by running the `mma.sync.aligned.m8n32k16.row.col.s32.s8.s8.s32` [PTX](https://en.wikipedia.org/wiki/Parallel_Thread_Execution) instruction, or calling `wmma::mma_sync` at the CUDA level. However, both approaches require careful management of data movement and layouts to maximize Tensor Core throughput. 

Thankfully, these lower level details are abstracted away by the cuBLASLt  `cublasLtMatmul`  and CUTLASS `device::Gemm` APIs, both of which support IMMA (integer matrix multiply accumulate).

## Available Solutions

> While integration with these APIs is currently not supported natively in PyTorch, there are other libraries available such as [**torch-int**](https://github.com/Guangxuan-Xiao/torch-int) (SmoothQuant) and [**bitsandbytes**](https://github.com/TimDettmers/bitsandbytes) (LLM.int8()) which expose Python bindings to the underlying cuBLASLt/CUTLASS calls. Microsoft's **ZeroQuant** also leverage CUTLASS, but wrappers for their INT8 kernels are notz open source.

Reviewer: ```suggestion
While integration with these APIs is currently not supported natively in PyTorch, there are other libraries available such as [**torch-int**](https://github.com/Guangxuan-Xiao/torch-int) (SmoothQuant) and [**bitsandbytes**](https://github.com/TimDettmers/bitsandbytes) (LLM.int8()) which expose Python bindings to the underlying cuBLASLt/CUTLASS calls. Microsoft's **ZeroQuant** also leverage CUTLASS, but wrappers for their INT8 kernels are not open source.
```

Although these libraries offer flexibility and easy integration, they don't provide performance gains and are consistently slower than FP16. This is due to prioritizing accuracy and memory savings or lacking efficient quantization implementations.

In contrast, fully-fledged inference frameworks such as **TensorRT (TRT)** and **FasterTransformer** do achieve performance gains. They also manage the complexity of fusing quant & dequant nodes with adjacent operators. This is appealing for common Transformer types such as BERT and GPT, for which they have been heavily optimised. However, the rigid assumptions made by these libraries make them less suitable for more unusual architectures. 

> Specifically, TRT rewrites each model using the TRT Network Definition API, and utilizes custom plugins (such as fused multi-headed attention). This is essential for peak performance, but this level of manual intervention means the benefits of a more generic model export + inference runtime are diminished. FasterTransformer's INT8 compatible models have been rewritten in C++ in order to leverage the best performance from cuBLASLt non-standard interleaved data layout. This limits the extensibility of existing INT8 model implementations for novel architectures.

Reviewer: ```suggestion
Specifically, whilst TRT supports generic ONNX models, to achieve peak performance in their BERT implementation they rewrite the model using the TRT Network Definition API, and utilize custom plugins (such as fused multi-headed attention). This level of manual intervention means the benefits of a more generic model export + inference runtime are diminished. FasterTransformer's INT8 compatible models have been rewritten in C++ in order to leverage the best performance by using the non-standard interleaved data layout required by cuBLASLt. This limits the extensibility of existing INT8 model implementations for novel architectures.
```

## Memory Layouts

As previously suggested, ensuring that input and weight matrices satisfy specific memory layout requirements is essential for INT8 GEMM performance . By default, all PyTorch operators expect a **row major** ordering for input and outputs tensors. Ideally, we'd use the same layout for our INT8 matmul to avoid conversion overhead.

Unfortunately, this is not the case with cuBLASLt, which operates on **column major** by default. The `cublasLtMatmul` API does support a row major input tensor with column major weight tensor (and we can transpose the weight tensor offline), but the output is returned in column major. In other words, input/weight/output = `ROW`/`COL`/`COL`. CUTLASS  goes further and supports `ROW`/`COL`/`ROW` out of the box, which makes it a great option for PyTorch integrations.

> While these options are already faster than FP16, maximizing performance requires that our input tensors are ordered in the exceptionally non-standard `COL32`/`CUBLASLT_ORDER_COL32_2R_4R4`/`COL32` layout.

Reviewer: maximum? or something else

`COL32` is an interleaved layout which can be interpreted as row-major ordered, but in blocks of 32 columns. CUTLASS supports this by specifying `cutlass::layout::ColumnMajorInterleaved<32>`.

> There also exists `CUBLASLT_ORDER_COL32_2R_4R4`, which is even more exotic! 

Reviewer: I am happy to keep this closer to the original, e.g. 

> CUBLASLT_ORDER_COL32_2R_4R4 is even more exotic, and is probably best explained visually... 

But breaking up the the diagram into a separate sentence as you have done

I don't think we need to say there also exists as we have referenced it a few lines before

#### Column Ampere (CUBLASLT_ORDER_COL32_2R_4R4)
![](_attachments/Pasted%20image%2020230329113458.png)

While `COL32` is the most performant layout, it comes with an associated cost of layout conversion. This may cancel out any gains from the reduced precision matmul. Therefore, we must decide to either:

> 1. Persist the data in the required format (à la FasterTransformer ).

Reviewer: ```suggestion
1. Persist the data in the required format (à la FasterTransformer).
```

Maybe add a link to the FT section here?

What if we require **per-channel** quantization? In this case, CUTLASS comes to the rescue by allowing the definition of a custom epilogue function. This is applied after the matrix multiplication, in a single fused kernel. The GEMM + epilogue definition is expanded to:

$$D=f_2(f_1(\alpha AB+\beta C, d)) $$
The epilogue format comes from `EpilogueWithBroadcast` which applies a [binary operation](https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/epilogue/thread/linear_combination_bias_elementwise.h#L215)`f1` between the matmul output and a column-wise broadcasted vector `d`, followed by an optional elementwise op `f2`.

`f1` might typically be a bias addition followed by an activation function (e.g. ReLU), but in our case we want it to be a multiplication with the dequantization scalar. The epilogue is then plugged into `gemm::device::GemmUniversalWithBroadcast`.

The throughput figures we measured are shown below:

|       Kernel       | Time (ms) | vs. FP16 |
|:------------------:|:---------:|:--------:|
| f16f16f16 (Torch)  |    600    |   1.0x   |
| i8i8i32 (CUTLASS) |    461    |   1.30x  |
| i8i8f16 (CUTLASS) |    438    |   1.37x  |

> Whilst there might not huge throughput improvements from FP16 output, there are other peformance benefits:

Reviewer: ```suggestion
Whilst there might not be huge throughput improvements from FP16 output, there are other performance benefits:
```

- 50% less data loaded in the next kernel (now FP16 instead of INT32)
- Avoid fusion of the dequantize operator with the next kernel
- Avoid loading the dequantization vector in the next kernel (which CUTLASS pipelines the loading of)

> [!TODO]
> Improve the last sentence

### Memory layout

Lastly, we examine the effect of memory layout on matmul performance:

|       Kernel       | Time (ms) | vs. FP16 |
|:------------------:|:---------:|:--------:|
| FP16 Row major (Torch)  |    600    |   1.0x   |
| INT8 Row major  (cuBLASLt) |    365    |   1.64x  |
| INT8 COL32 (cuBLASLt) |    308    |   1.95x  |

As expected, COL32 is most performant. 

# Part III: FP8 & The Future of 8-bit Quantization

The arrival of Nvidia's Hopper/Lovelace architectures brings support for a new floating point datatype - FP8. This is available in two formats:

- **E5M2** - 5 exponent bits and 2 mantissa bits - larger dynamic range
- **E4M3** - 4 exponent bits and 3 mantissa bits - higher precision

> Choosing an FP8 quantization format could have both accuracy and performance benefits.

Reviewer: ```suggestion
Choosing an FP8 quantization format can have both accuracy and performance benefits.
```
